### ISSUE TITLE
[ENHANCEMENTS] Cluster capability?

### ISSUE THREAD

**Issue by sahara101** [tobychui/zoraxy#183]
**Is your feature request related to a problem? Please describe.**

If the reverse proxy host goes down - all proxies stop working

**Describe the solution you'd like**

Possibility to have a cluster of zoraxy, for HA.

Would this be possible?

Thanks!

**Comment by tobychui** [comment]
Well I mean you can always just setup two Zoraxy node with the same configs?
Or you want config auto sync between two nodes? Without DDNS it is not possible to do auto-swap or something like that so I am not pretty sure what you meant by cluster.

**Comment by sahara101** [comment]
I would only need to setup keepalived and assign a VIP. Yes, I do not know how I would sync the configs from master to slave. For example Pinole uses gravity-sync.

**Comment by Tacioandrade** [comment]
> Well I mean you can always just setup two Zoraxy node with the same configs? Or you want config auto sync between two nodes? Without DDNS it is not possible to do auto-swap or something like that so I am not pretty sure what you meant by cluster.

In this case, I believe it would be a way to synchronize the settings between 2 or more zoraxy, both the generated certificates, the proxy, users, etc.
In this case, the HA part for the application would be done by a DDNS or some tool at a higher level, such as the network gateway itself, but so as not to have to sync the configurations via rsync or something like that, sync compatibility of the Application-level data would be interesting.

**Comment by CybergNetwork** [comment]
I know it's an old post but the Cluster capability has no business with Zoraxy. It comes from your own deployed infrastructure. That's your infrastructure that makes your containers HA.

**Comment by Tacioandrade** [comment]
> I know it's an old post but the Cluster capability has no business with Zoraxy. It comes from your own deployed infrastructure. That's your infrastructure that makes your containers HA.

The objective of the Cluster would be to have a replica of zoraxy on another host with another IP, for example one in us-east1a and another in us-east1b of AWS for example and in case of a datacenter failure, the application will continue to be functional, running through the second host.

**Comment by CybergNetwork** [comment]
If you want a replica, you need a backend network implemented to sync the databases asynchronous. In additon you still need a load balancer that will look at the health status for each Zoraxy instance. If you wanna use replica and you can't implement replica yourself, how are you going to troubleshoot it if it goes down? Search Google or ask your question to an AI. Your next step is to learn, not to implement.

**Comment by Tacioandrade** [comment]
> If you want a replica, you need a backend network implemented to sync the databases asynchronous. In additon you still need a load balancer that will look at the health status for each Zoraxy instance. If you wanna use replica and you can't implement replica yourself, how are you going to troubleshoot it if it goes down? Search Google or ask your question to an AI. Your next step is to learn, not to implement.

Yes, you can actually call it a replica! In this case, I called it a Cluster because it is what other tools like nxfilter, pfSense, etc. call it in environments where changes are implemented on host1 and it replicates the changes to the other hosts that are ready to accept connections, whether they are multi-master or master/slave clusters.

In the case of ngixn proxy manager, a friend implemented the replication part behind the scenes, syncing all the configuration files and certificates from the main host to the others, but something native directly through the tool could be interesting, if the developers thought it feasible to implement.

**Comment by IoannisTM** [comment]
hi @Tacioandrade  have you figured out a way to do HA for zoraxy?
I am also looking at this without too much luck

**Comment by Tacioandrade** [comment]
> hi [@Tacioandrade](https://github.com/Tacioandrade) have you figured out a way to do HA for zoraxy? I am also looking at this without too much luck

I really liked the project, but I ended up choosing the nginx proxy manager for my environment due to its peculiarities.

For that reason, I didn't pursue it further. However, with the nginx proxy manager, I end up doing the replication via script. During the night, I gzip the nginx proxy manager volume, send it via scp to my main server, and 15 minutes later, my destination server runs a cron job that pulls this file, generates a backup of the current configuration, overwrites it with the new files, and runs `docker compose down && docker compose up -d`, updating the project to the correct version.

It's not the best option, but it's the best I could do and it has worked for me.

**Comment by tobychui** [comment]
@Tacioandrade You can do the same with Zoraxy, just setup a sync services for `sys.db` and all files under the `./conf` folder and do the sync nightly. But since you already have a working setup, it is better off not touch that for now.

**Comment by Tacioandrade** [comment]
> [@Tacioandrade](https://github.com/Tacioandrade) You can do the same with Zoraxy, just setup a sync services for `sys.db` and all files under the `./conf` folder and do the sync nightly. But since you already have a working setup, it is better off not touch that for now.

Great point my friend! In this case, I only replaced zoraxy with the nginx proxy manager, because when I tested it as the TSPlus reverse proxy, for some reason the delay ended up being greater, which prevented it from being put into production.

Since that didn't work, I didn't look for this sync functionality, however, I'm happy to know that we can do the same thing with it, because if I need it in the future, I already know how to implement it. =)

**Comment by saineela** [comment]
setup uptime kuma, track zoraxy and when not responsive, trigger a webhook with a custom python script on your machine to turn on the second zoraxy instance!
